<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Course material (lectures and workshops) from Stephanie Hicks">

<title>CSHL Computational Genomics 2023 - 6&nbsp; Overview of single-cell analysis in R/Bioconductor</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./exercises-04.html" rel="next">
<link href="./intro-to-bioc.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./intro-to-single-cell.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Overview of single-cell analysis in R/Bioconductor</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">CSHL Computational Genomics 2023</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro-to-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Group work</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./reproducible-res.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Reproducible Research</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Group work</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Tidyverse and data viz</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Workshop</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro-to-bioc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Bioconductor</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro-to-single-cell.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Overview of single-cell analysis in R/Bioconductor</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Workshop</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#part-1" id="toc-part-1" class="nav-link active" data-scroll-target="#part-1"><span class="header-section-number">6.1</span> Part 1</a>
  <ul class="collapse">
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives"><span class="header-section-number">6.1.1</span> Learning objectives</a></li>
  <li><a href="#materials" id="toc-materials" class="nav-link" data-scroll-target="#materials"><span class="header-section-number">6.1.2</span> Materials</a></li>
  </ul></li>
  <li><a href="#part-2" id="toc-part-2" class="nav-link" data-scroll-target="#part-2"><span class="header-section-number">6.2</span> Part 2</a>
  <ul class="collapse">
  <li><a href="#learning-objectives-1" id="toc-learning-objectives-1" class="nav-link" data-scroll-target="#learning-objectives-1"><span class="header-section-number">6.2.1</span> Learning objectives</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview"><span class="header-section-number">6.2.2</span> Overview</a></li>
  <li><a href="#creating-a-count-matrix" id="toc-creating-a-count-matrix" class="nav-link" data-scroll-target="#creating-a-count-matrix"><span class="header-section-number">6.2.3</span> Creating a count matrix</a></li>
  <li><a href="#singlecellexperiment-class" id="toc-singlecellexperiment-class" class="nav-link" data-scroll-target="#singlecellexperiment-class"><span class="header-section-number">6.2.4</span> <code>SingleCellExperiment</code> Class</a></li>
  </ul></li>
  <li><a href="#part-3" id="toc-part-3" class="nav-link" data-scroll-target="#part-3"><span class="header-section-number">6.3</span> Part 3</a>
  <ul class="collapse">
  <li><a href="#learning-objectives-2" id="toc-learning-objectives-2" class="nav-link" data-scroll-target="#learning-objectives-2"><span class="header-section-number">6.3.1</span> Learning objectives</a></li>
  <li><a href="#overview-1" id="toc-overview-1" class="nav-link" data-scroll-target="#overview-1"><span class="header-section-number">6.3.2</span> Overview</a></li>
  <li><a href="#quick-start-simple" id="toc-quick-start-simple" class="nav-link" data-scroll-target="#quick-start-simple"><span class="header-section-number">6.3.3</span> Quick start (simple)</a></li>
  <li><a href="#quick-start-multiple-batches" id="toc-quick-start-multiple-batches" class="nav-link" data-scroll-target="#quick-start-multiple-batches"><span class="header-section-number">6.3.4</span> Quick start (multiple batches)</a></li>
  </ul></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="header-section-number">6.4</span> Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Overview of single-cell analysis in R/Bioconductor</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="part-1" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="part-1"><span class="header-section-number">6.1</span> Part 1</h2>
<section id="learning-objectives" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">6.1.1</span> Learning objectives</h3>
<ol type="1">
<li>Understand how count matrices are created from single-cell experimental platforms and protocols</li>
<li>Recognize which basic principles and concepts were transferred from bulk to single-cell data analyses</li>
<li>Understand the key differences between bulk and single-cell data</li>
<li>Define what is a “Unique Molecular Identifier”</li>
<li>Define multiplexing (and demultiplexing)</li>
<li>Overview of basics in single-cell data analyses including quality control, normalization, feature selection, dimensionality reduction, and clustering</li>
</ol>
</section>
<section id="materials" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="materials"><span class="header-section-number">6.1.2</span> Materials</h3>
<p>We will go through the slides available here:</p>
<ul>
<li><a href="https://docs.google.com/presentation/d/1XHfN-5NCXtu1P4reotrckGplPXJBTlQraFpd8ViPf1g/edit?usp=sharing" class="uri">https://docs.google.com/presentation/d/1XHfN-5NCXtu1P4reotrckGplPXJBTlQraFpd8ViPf1g/edit?usp=sharing</a></li>
</ul>
</section>
</section>
<section id="part-2" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="part-2"><span class="header-section-number">6.2</span> Part 2</h2>
<section id="learning-objectives-1" class="level3" data-number="6.2.1">
<h3 data-number="6.2.1" class="anchored" data-anchor-id="learning-objectives-1"><span class="header-section-number">6.2.1</span> Learning objectives</h3>
<ol type="1">
<li>Be able to create a single-cell count matrix and read it into R</li>
<li>Recognize and define the <code>SingleCellExperiment</code> S4 class in R/Bioconductor to store single-cell data</li>
</ol>
</section>
<section id="overview" class="level3" data-number="6.2.2">
<h3 data-number="6.2.2" class="anchored" data-anchor-id="overview"><span class="header-section-number">6.2.2</span> Overview</h3>
<p>NGS data from scRNA-seq experiments must be converted into a matrix of expression values.</p>
<p>This is <strong>usually a count matrix</strong> containing the number of reads (or UMIs) mapped to each gene (row) in each cell (column). Once this quantification is complete, we can proceed with our downstream statistical analyses in R.</p>
<p>Constructing a count matrix from raw scRNA-seq data requires some thought as the term “single-cell RNA-seq” encompasses a variety of different experimental protocols. This includes</p>
<ul>
<li><strong>droplet-based protocols</strong> like 10X Genomics, inDrop and Drop-seq</li>
<li><strong>plate-based protocols</strong> with UMIs like CEL-seq(2) and MARS-seq</li>
<li><strong>plate-based protocols</strong> with reads (mostly Smart-seq2)</li>
<li>others like sciRNA-seq, etc</li>
</ul>
<p>Each approach requires a different processing pipeline to deal with cell demultiplexing and UMI deduplication (if applicable). Here, we will briefly describe some of the methods used to generate a count matrix and read it into R.</p>
</section>
<section id="creating-a-count-matrix" class="level3" data-number="6.2.3">
<h3 data-number="6.2.3" class="anchored" data-anchor-id="creating-a-count-matrix"><span class="header-section-number">6.2.3</span> Creating a count matrix</h3>
<p>As mentioned above, the exact procedure for quantifying expression depends on the technology involved:</p>
<ul>
<li>For <strong>10X Genomics data</strong>, the <code>Cellranger</code> software suite (Zheng et al.&nbsp;2017) provides a custom pipeline to obtain a count matrix. This uses STAR to align reads to the reference genome and then counts the number of unique UMIs mapped to each gene.</li>
<li>Alternatively, pseudo-alignment methods such as <a href="https://doi.org/10.1186/s13059-019-1670-y"><code>alevin</code></a> (Srivastava et al.&nbsp;2019) can be used to obtain a count matrix from the same data. This avoids the need for explicit alignment, which reduces the compute time and memory usage.</li>
<li>For other <strong>highly multiplexed protocols</strong>, the <a href="http://bioconductor.org/packages/scPipe/"><code>scPipe</code></a> package provides a more general pipeline for processing scRNA-seq data. This uses the Rsubread aligner to align reads and then counts reads or UMIs per gene.</li>
<li>For <strong>CEL-seq or CEL-seq2 data</strong>, the <code>scruff</code> package provides a dedicated pipeline for quantification.</li>
<li>For <strong>read-based protocols</strong>, we can generally re-use the same pipelines for processing bulk RNA-seq data (e.g.&nbsp;Subread, RSEM, salmon)</li>
<li>For <strong>any data involving spike-in transcripts</strong>, the spike-in sequences should be included as part of the reference genome during alignment and quantification.</li>
</ul>
<p>In all cases, the <strong>identity of the genes</strong> in the count matrix should be <strong>defined with standard identifiers from Ensembl or Entrez</strong>. These provide an unambiguous mapping between each row of the matrix and the corresponding gene.</p>
<p>In contrast, <strong>a single gene symbol</strong> may be used by multiple loci, or the mapping between symbols and genes <strong>may change over time</strong>, e.g., if the gene is renamed.</p>
<p>This makes it difficult to re-use the count matrix as we cannot be confident in the meaning of the symbols.</p>
<p>(Of course, <strong>identifiers can be easily converted to gene symbols</strong> later on in the analysis. This is the recommended approach as it allows us to document how the conversion was performed and to backtrack to the stable identifiers if the symbols are ambiguous.)</p>
</section>
<section id="singlecellexperiment-class" class="level3" data-number="6.2.4">
<h3 data-number="6.2.4" class="anchored" data-anchor-id="singlecellexperiment-class"><span class="header-section-number">6.2.4</span> <code>SingleCellExperiment</code> Class</h3>
<p>One of the <strong>main strengths</strong> of the Bioconductor project lies in the use of <strong>a common data infrastructure</strong> that powers interoperability across packages.</p>
<p>Users should be able to analyze their data using functions from different Bioconductor packages without the need to convert between formats.</p>
<p>To this end, the <code>SingleCellExperiment</code> class (from the <code>SingleCellExperiment</code> package) serves as the common currency for data exchange across 70+ single-cell-related Bioconductor packages.</p>
<p>This class implements a data structure that stores all aspects of our single-cell data - gene-by-cell expression data, per-cell metadata and per-gene annotation - and manipulate them in a synchronized manner.</p>
<div class="cell" data-layout-align="center" data-show="true" data-fig.caption="Overview of the structure of the `SingleCellExperiment` class. Each row of the assays corresponds to a row of the `rowData` (pink shading), while each column of the assays corresponds to a column of the `colData` and `reducedDims` (yellow shading).">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figures/SingleCellExperiment.png" class="img-fluid figure-img" width="780"></p>
</figure>
</div>
</div>
</div>
<p>[Amezquita et al.&nbsp;2019 (<a href="https://doi.org/10.1101/590562" class="uri">https://doi.org/10.1101/590562</a>)]</p>
<ul>
<li>Each piece of (meta)data in the <code>SingleCellExperiment</code> is <strong>represented by a separate “slot”</strong>.</li>
</ul>
<p>(This terminology comes from the S4 class system, but that’s not important right now).</p>
<p>If we imagine the <code>SingleCellExperiment</code> object to be a cargo ship, the <strong>slots can be thought of as individual cargo boxes with different contents</strong>, e.g., certain slots expect numeric matrices whereas others may expect data frames.</p>
<p>If you want to know more about the available slots, their expected formats, and how we can interact with them, check out this <a href="https://bioconductor.org/books/release/OSCA.intro/the-singlecellexperiment-class.html">chapter</a>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<code>SingleCellExperiment</code> Example
</div>
</div>
<div class="callout-body-container callout-body">
<p>Let’s show you what a <code>SingleCellExperiment</code> (or <code>sce</code> for short) looks like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>sce</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 20006 3005 
metadata(0):
assays(1): counts
rownames(20006): Tspan12 Tshz1 ... mt-Rnr1 mt-Nd4l
rowData names(1): featureType
colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12
  1772058148_F03
colData names(10): tissue group # ... level1class level2class
reducedDimNames(0):
mainExpName: endogenous
altExpNames(2): ERCC repeat</code></pre>
</div>
</div>
<p>This <code>SingleCellExperiment</code> object has 20006 genes and 3005 cells.</p>
<p>We can pull out the counts matrix with the <code>counts()</code> function and the corresponding <code>rowData()</code> and <code>colData()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(sce)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         1772071015_C02 1772071017_G12 1772071017_A05 1772071014_B06
Tspan12               0              0              0              3
Tshz1                 3              1              0              2
Fnbp1l                3              1              6              4
Adamts15              0              0              0              0
Cldn12                1              1              1              0
         1772067065_H06
Tspan12               0
Tshz1                 2
Fnbp1l                1
Adamts15              0
Cldn12                0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 20006 rows and 1 column
         featureType
         &lt;character&gt;
Tspan12   endogenous
Tshz1     endogenous
Fnbp1l    endogenous
Adamts15  endogenous
Cldn12    endogenous
...              ...
mt-Co2          mito
mt-Co1          mito
mt-Rnr2         mito
mt-Rnr1         mito
mt-Nd4l         mito</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 3005 rows and 10 columns
                       tissue   group # total mRNA mol      well       sex
                  &lt;character&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
1772071015_C02       sscortex         1           1221         3         3
1772071017_G12       sscortex         1           1231        95         1
1772071017_A05       sscortex         1           1652        27         1
1772071014_B06       sscortex         1           1696        37         3
1772067065_H06       sscortex         1           1219        43         3
...                       ...       ...            ...       ...       ...
1772067059_B04 ca1hippocampus         9           1997        19         1
1772066097_D04 ca1hippocampus         9           1415        21         1
1772063068_D01       sscortex         9           1876        34         3
1772066098_A12 ca1hippocampus         9           1546        88         1
1772058148_F03       sscortex         9           1970        15         3
                     age  diameter        cell_id       level1class level2class
               &lt;numeric&gt; &lt;numeric&gt;    &lt;character&gt;       &lt;character&gt; &lt;character&gt;
1772071015_C02         2         1 1772071015_C02      interneurons       Int10
1772071017_G12         1       353 1772071017_G12      interneurons       Int10
1772071017_A05         1        13 1772071017_A05      interneurons        Int6
1772071014_B06         2        19 1772071014_B06      interneurons       Int10
1772067065_H06         6        12 1772067065_H06      interneurons        Int9
...                  ...       ...            ...               ...         ...
1772067059_B04         4       382 1772067059_B04 endothelial-mural       Peric
1772066097_D04         7        12 1772066097_D04 endothelial-mural        Vsmc
1772063068_D01         7       268 1772063068_D01 endothelial-mural        Vsmc
1772066098_A12         7       324 1772066098_A12 endothelial-mural        Vsmc
1772058148_F03         7         6 1772058148_F03 endothelial-mural        Vsmc</code></pre>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="part-3" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="part-3"><span class="header-section-number">6.3</span> Part 3</h2>
<section id="learning-objectives-2" class="level3" data-number="6.3.1">
<h3 data-number="6.3.1" class="anchored" data-anchor-id="learning-objectives-2"><span class="header-section-number">6.3.1</span> Learning objectives</h3>
<ol type="1">
<li>Be able to describe a standard workflow for analyzing single-cell data</li>
<li>Be able to run code for a standard workflow starting from loading a <code>SingleCellExperiment</code> in R and identifying clusters.</li>
</ol>
</section>
<section id="overview-1" class="level3" data-number="6.3.2">
<h3 data-number="6.3.2" class="anchored" data-anchor-id="overview-1"><span class="header-section-number">6.3.2</span> Overview</h3>
<p>Here, we provide an overview of the framework of a typical scRNA-seq analysis workflow:</p>
<div class="cell" data-layout-align="center" data-show="true" data-fig.caption="Schematic of a typical scRNA-seq analysis workflow. Each stage (separated by dashed lines) consists of a number of specific steps, many of which operate on and modify a SingleCellExperiment instance.">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figures/workflow.png" class="img-fluid figure-img" width="780"></p>
</figure>
</div>
</div>
</div>
<p>In the simplest case, the workflow has the following form:</p>
<ol type="1">
<li>We compute <strong>quality control metrics</strong> to remove low-quality cells that would interfere with downstream analyses. These cells may have been damaged during processing or may not have been fully captured by the sequencing protocol. Common metrics includes the total counts per cell, the proportion of spike-in or mitochondrial reads and the number of detected features.</li>
<li>We convert the counts into <strong>normalized expression values</strong> to eliminate cell-specific biases (e.g., in capture efficiency). This allows us to perform explicit comparisons across cells in downstream steps like clustering. We also apply a transformation, typically log, to adjust for the mean-variance relationship.</li>
<li>We perform <strong>feature selection to pick a subset of interesting features</strong> for downstream analysis. This is done by modelling the variance across cells for each gene and retaining genes that are highly variable. The aim is to reduce computational overhead and noise from uninteresting genes.</li>
<li>We apply <strong>dimensionality reduction to compact the data</strong> and further reduce noise. Principal components analysis is typically used to obtain an initial low-rank representation for more computational work, followed by more aggressive methods like <span class="math inline">\(t\)</span>-stochastic neighbor embedding for visualization purposes.</li>
<li>We <strong>cluster cells into groups</strong> according to similarities in their (normalized) expression profiles. This aims to obtain groupings that serve as empirical proxies for distinct biological states. We typically interpret these groupings by identifying differentially expressed marker genes between clusters.</li>
</ol>
</section>
<section id="quick-start-simple" class="level3" data-number="6.3.3">
<h3 data-number="6.3.3" class="anchored" data-anchor-id="quick-start-simple"><span class="header-section-number">6.3.3</span> Quick start (simple)</h3>
<p>Here, we use the a droplet-based retina dataset from Macosko et al.&nbsp;(2015), provided in the <code>scRNAseq</code> package. This starts from a count matrix and finishes with clusters in preparation for biological interpretation. We also demonstrate how to identify differentially expressed genes between the clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scRNAseq)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">MacoskoRetinaData</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?scRNAseq and browseVignettes('scRNAseq') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?scRNAseq and browseVignettes('scRNAseq') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Quality control (using mitochondrial genes).</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>is.mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"^MT-"</span>, <span class="fu">rownames</span>(sce))</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>qcstats <span class="ot">&lt;-</span> <span class="fu">perCellQCMetrics</span>(sce, <span class="at">subsets=</span><span class="fu">list</span>(<span class="at">Mito=</span>is.mito))</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>filtered <span class="ot">&lt;-</span> <span class="fu">quickPerCellQC</span>(qcstats, <span class="at">percent_subsets=</span><span class="st">"subsets_Mito_percent"</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="sc">!</span>filtered<span class="sc">$</span>discard]</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalization.</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Feature selection.</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>hvg <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop=</span><span class="fl">0.1</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA.</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">ncomponents=</span><span class="dv">25</span>, <span class="at">subset_row=</span>hvg)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Clustering.</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bluster)</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a><span class="fu">colLabels</span>(sce) <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce, <span class="at">use.dimred=</span><span class="st">'PCA'</span>,</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM=</span><span class="fu">NNGraphParam</span>(<span class="at">cluster.fun=</span><span class="st">"louvain"</span>))    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualization.</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(sce, <span class="at">dimred =</span> <span class="st">'PCA'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotUMAP</span>(sce, <span class="at">colour_by=</span><span class="st">"label"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="intro-to-single-cell_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">UMAP plot of the retina dataset, where each point is a cell and is colored by the assigned cluster identity.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Marker detection (between pairs of groups)</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(sce, <span class="at">test.type=</span><span class="st">"wilcox"</span>, <span class="at">direction=</span><span class="st">"up"</span>, <span class="at">lfc=</span><span class="dv">1</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 13</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>markers[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 24658 rows and 16 columns
                    Top      p.value          FDR summary.AUC     AUC.2
              &lt;integer&gt;    &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;
MEG3                  1  0.00000e+00  0.00000e+00    0.867306  0.857315
TUBA1A                1  3.58784e-81  5.89793e-78    0.609195  0.479862
SNHG11                1  0.00000e+00  0.00000e+00    0.737419  0.747343
SYT1                  2 3.51850e-268 1.08449e-264    0.786614  0.454677
CALM1                 2  0.00000e+00  0.00000e+00    0.812019  0.665619
...                 ...          ...          ...         ...       ...
VSIG1             24654            1            1           0         0
GM16390           24655            1            1           0         0
GM25207           24656            1            1           0         0
1110059M19RIK     24657            1            1           0         0
GM20861           24658            1            1           0         0
                  AUC.3     AUC.4     AUC.5     AUC.6     AUC.7     AUC.8
              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
MEG3          0.6814265  0.463403  0.806255  0.478814 0.1822596  0.888640
TUBA1A        0.1162662  0.315768  0.422655  0.289249 0.4529957  0.541010
SNHG11        0.0822847  0.562474  0.741261  0.703891 0.0228628  0.742098
SYT1          0.5354274  0.317921  0.786614  0.307464 0.0591646  0.549534
CALM1         0.1535742  0.370682  0.715928  0.425258 0.0105809  0.791806
...                 ...       ...       ...       ...       ...       ...
VSIG1                 0         0         0         0         0         0
GM16390               0         0         0         0         0         0
GM25207               0         0         0         0         0         0
1110059M19RIK         0         0         0         0         0         0
GM20861               0         0         0         0         0         0
                  AUC.9     AUC.10    AUC.11    AUC.12    AUC.13
              &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
MEG3           0.850705 0.48570145  0.866916  0.873796  0.867306
TUBA1A         0.509171 0.33506617  0.552997  0.566389  0.609195
SNHG11         0.745361 0.74450392  0.737413  0.744927  0.737419
SYT1           0.823158 0.39904104  0.575948  0.584176  0.633476
CALM1          0.444978 0.00587859  0.749949  0.834002  0.812019
...                 ...        ...       ...       ...       ...
VSIG1                 0          0         0         0         0
GM16390               0          0         0         0         0
GM25207               0          0         0         0         0
1110059M19RIK         0          0         0         0         0
GM20861               0          0         0         0         0</code></pre>
</div>
</div>
</section>
<section id="quick-start-multiple-batches" class="level3" data-number="6.3.4">
<h3 data-number="6.3.4" class="anchored" data-anchor-id="quick-start-multiple-batches"><span class="header-section-number">6.3.4</span> Quick start (multiple batches)</h3>
<p>If you could like to see Quick Start for scRNA-seq data with multiple batches, check out this workflow:</p>
<ul>
<li><a href="https://bioconductor.org/books/release/OSCA.intro/analysis-overview.html#quick-start-multiple-batches" class="uri">https://bioconductor.org/books/release/OSCA.intro/analysis-overview.html#quick-start-multiple-batches</a></li>
</ul>
</section>
</section>
<section id="session-info" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="session-info"><span class="header-section-number">6.4</span> Session Info</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.1.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] bluster_1.10.0              scran_1.28.2               
 [3] scRNAseq_2.14.0             scater_1.28.0              
 [5] ggplot2_3.4.4               scuttle_1.9.4              
 [7] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2
 [9] Biobase_2.60.0              GenomicRanges_1.52.1       
[11] GenomeInfoDb_1.36.4         IRanges_2.34.1             
[13] S4Vectors_0.38.2            BiocGenerics_0.46.0        
[15] MatrixGenerics_1.12.3       matrixStats_1.0.0          

loaded via a namespace (and not attached):
  [1] rstudioapi_0.15.0             jsonlite_1.8.7               
  [3] magrittr_2.0.3                ggbeeswarm_0.7.2             
  [5] GenomicFeatures_1.52.2        farver_2.1.1                 
  [7] rmarkdown_2.25                BiocIO_1.10.0                
  [9] zlibbioc_1.46.0               vctrs_0.6.4                  
 [11] memoise_2.0.1                 Rsamtools_2.16.0             
 [13] DelayedMatrixStats_1.22.6     RCurl_1.98-1.12              
 [15] htmltools_0.5.6.1             S4Arrays_1.0.6               
 [17] progress_1.2.2                AnnotationHub_3.8.0          
 [19] curl_5.1.0                    BiocNeighbors_1.18.0         
 [21] htmlwidgets_1.6.2             cachem_1.0.8                 
 [23] GenomicAlignments_1.36.0      igraph_1.5.1                 
 [25] mime_0.12                     lifecycle_1.0.4              
 [27] pkgconfig_2.0.3               rsvd_1.0.5                   
 [29] Matrix_1.6-1.1                R6_2.5.1                     
 [31] fastmap_1.1.1                 GenomeInfoDbData_1.2.10      
 [33] shiny_1.7.5.1                 digest_0.6.33                
 [35] colorspace_2.1-0              AnnotationDbi_1.62.2         
 [37] dqrng_0.3.1                   irlba_2.3.5.1                
 [39] ExperimentHub_2.8.1           RSQLite_2.3.1                
 [41] beachmat_2.16.0               labeling_0.4.3               
 [43] filelock_1.0.2                fansi_1.0.5                  
 [45] httr_1.4.7                    abind_1.4-5                  
 [47] compiler_4.3.1                bit64_4.0.5                  
 [49] withr_2.5.2                   BiocParallel_1.34.2          
 [51] viridis_0.6.4                 DBI_1.1.3                    
 [53] biomaRt_2.56.1                rappdirs_0.3.3               
 [55] DelayedArray_0.26.7           rjson_0.2.21                 
 [57] tools_4.3.1                   vipor_0.4.5                  
 [59] beeswarm_0.4.0                interactiveDisplayBase_1.38.0
 [61] httpuv_1.6.12                 glue_1.6.2                   
 [63] restfulr_0.0.15               promises_1.2.1               
 [65] grid_4.3.1                    cluster_2.1.4                
 [67] generics_0.1.3                gtable_0.3.4                 
 [69] ensembldb_2.24.1              hms_1.1.3                    
 [71] metapod_1.7.0                 BiocSingular_1.16.0          
 [73] ScaledMatrix_1.8.1            xml2_1.3.5                   
 [75] utf8_1.2.4                    XVector_0.40.0               
 [77] RcppAnnoy_0.0.21              ggrepel_0.9.4                
 [79] BiocVersion_3.17.1            pillar_1.9.0                 
 [81] stringr_1.5.1                 limma_3.56.2                 
 [83] later_1.3.1                   dplyr_1.1.4                  
 [85] BiocFileCache_2.8.0           lattice_0.22-5               
 [87] rtracklayer_1.60.1            bit_4.0.5                    
 [89] tidyselect_1.2.0              locfit_1.5-9.8               
 [91] Biostrings_2.68.1             knitr_1.44                   
 [93] gridExtra_2.3                 ProtGenerics_1.32.0          
 [95] edgeR_3.42.4                  xfun_0.40                    
 [97] statmod_1.5.0                 stringi_1.8.2                
 [99] lazyeval_0.2.2                yaml_2.3.7                   
[101] evaluate_0.22                 codetools_0.2-19             
[103] tibble_3.2.1                  BiocManager_1.30.22          
[105] cli_3.6.1                     uwot_0.1.16                  
[107] xtable_1.8-4                  munsell_0.5.0                
[109] Rcpp_1.0.11                   dbplyr_2.3.4                 
[111] png_0.1-8                     XML_3.99-0.15                
[113] parallel_4.3.1                ellipsis_0.3.2               
[115] blob_1.2.4                    prettyunits_1.2.0            
[117] AnnotationFilter_1.24.0       sparseMatrixStats_1.12.2     
[119] bitops_1.0-7                  viridisLite_0.4.2            
[121] scales_1.2.1                  purrr_1.0.2                  
[123] crayon_1.5.2                  rlang_1.1.2                  
[125] KEGGREST_1.40.1              </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./intro-to-bioc.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Bioconductor</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./exercises-04.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Workshop</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>